main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -43,29 +49,559 @@ object Lexer {
+      case _ if keyword("deref") => TokenKind.KeywordDeref
+      case _ if keyword("else") => TokenKind.KeywordElse
+      case _ if keyword("enum") => TokenKind.KeywordEnum
+      case _ if keyword("false") => TokenKind.KeywordFalse
+      case _ if keyword("fix") => TokenKind.KeywordFix
+      case _ if keyword("force") => TokenKind.KeywordForce
+      case _ if keyword("if") => TokenKind.KeywordIf
+      case _ if keyword("import") => TokenKind.KeywordImport
+      case _ if keyword("inline") => TokenKind.KeywordInline
+      case _ if keyword("instance") => TokenKind.KeywordInstance
+      case _ if keyword("into") => TokenKind.KeywordInto
+      case _ if keyword("law") => TokenKind.KeywordLaw
+      case _ if keyword("lawful") => TokenKind.KeywordLawful
+      case _ if keyword("lazy") => TokenKind.KeywordLazy
+      case _ if keyword("let") => TokenKind.KeywordLet
+      case _ if keyword("match") => TokenKind.KeywordMatch
+      case _ if keyword("typematch") => TokenKind.KeywordTypeMatch
+      case _ if keyword("null") => TokenKind.KeywordNull
+      case _ if keyword("override") => TokenKind.KeywordOverride
+      case _ if keyword("pub") => TokenKind.KeywordPub
+      case _ if keyword("ref") => TokenKind.KeywordRef
+      case _ if keyword("region") => TokenKind.KeywordRegion
+      case _ if keyword("sealed") => TokenKind.KeywordSealed
+      case _ if keyword("spawn") => TokenKind.KeywordSpawn
+      case _ if keyword("Static") => TokenKind.KeywordStatic
+      case _ if keyword("true") => TokenKind.KeywordTrue
+      case _ if keyword("type") => TokenKind.KeywordType
+      case _ if keyword("use") => TokenKind.KeywordUse
+      case _ if keyword("where") => TokenKind.KeywordWhere
+      case _ if keyword("with") => TokenKind.KeywordWith
+      case _ if keyword("discard") => TokenKind.KeywordDiscard
+      case _ if keyword("par") => TokenKind.KeywordPar
+      case _ if keyword("yield") => TokenKind.KeywordYield
+      case _ if isMathNameChar(c) => mathName()
+      case _ if isGreekNameChar(c) => greekName()
+
+      // User defined operators
+      case _ if validUserOpTokens.contains(c) => {
+        val p = peek()
+        if (validUserOpTokens.contains(p)) {
+          userDefinedOp()
+        } else if (c == '-' && p.isDigit) {
+          number() // negative numbers
+        } else if (c == '$' && p.isLetter) {
+          builtIn()
+        } else {
+          validUserOpTokens.apply(c)
+        }
+      }
+      case c if c.isLetter => name(c.isUpper)
+      case c if c.isDigit => number()
+      case '\"' => string()
+      case '\'' => char()
+      case '`' => infixFunction()
+      case _ => TokenKind.Err(TokenErrorKind.UnexpectedChar)
+    }
+
+    addToken(kind)
+  }
+
+  // Adds a token by consuming the characters between start and current
+  private def addToken(k: TokenKind)(implicit s: State): Unit = {
+    val t = s.src.data.slice(s.start.offset, s.current.offset).mkString("")
```

Comment:
This means that we create a new string for every token. This is probably a bit expensive. For example, we will now have X objects representing "def" floating around. Ideas for solutions?

Reply:
A very simplistic solution seems to be to just make `Token.text` optional. Then we could retain the text only where it is not derivable from `TokenKind`. What I don't like about that solution is that we don't encode what `TokenKind`s one should expect text on in the type system, so the programmer should "just know".

One way tackle that would be to make `Token.text` private optional and then define a `Token.get_text()` that matches on its own kind to produce its text. Then the details would be more neatly hidden.

What do you think?

Reply:
What about just storing the offset and a pointer to the entire string? Then we can compute the string by demand.

Reply:
Yep, I will try and do it like that